# Overview
### Dependencies
Ensure the following R packages are installed before running the script:
```
install.packages("data.table")
install.packages("stringr")
install.packages("susieR")
install.packages("coloc")
```
## Finemapping
### Usage
 ```
Rscript finemapping_analysis.R <variant> <summ_file> <ld_file> <lower> <upper> <chromosome>
 ```
### Arguments
 - ```variant```: SNP ID of the variant of interest
 - ```summ_file```: Path to the GWAS summary statistics file
 - ```ld_file```: Path to the LD matrix file (generated by plink)
 - ```lower```: Lower bound (bp) for the region to analyze
 - ```upper```: Upper bound (bp) for the region to analyze
 - ```chromosome```: Chromosome number of the variant
#### summ_file should contain the following columns:
 - ```SNP```: SNP ID in chr:pos:ref:alt format
 - ```allele1```: Effect allele (alt)
 - ```allele2```: Other allele (ref)
 - ```b```: Effect size estimate for effect allele
 - ```se```: Standard error
 - ```N```: Sample size
### Output
The script generates credible sets for the variant of interest, saved with suffix```_credible_set_susieR_results.txt```
## Colocalization
This script performs credible set extraction and colocalization analysis for a variant of interest using GWAS and eQTL summary statistics. The results include credible sets, colocalization probabilities, and identified variants.
### Usage
```
Rscript coloc_analysis.R <variant> <summ_file> <ld_file> <lower> <upper> <chromosome> <coloc_prob>
 ```
### Arguments
 - ```variant```: SNP ID of the variant of interest
 - ```summ_file```: Path to the GWAS summary statistics file
 - ```ld_file```: Path to the LD matrix file (generated by plink)
 - ```lower```: Lower bound (bp) for the region to analyze
 - ```upper```: Upper bound (bp) for the region to analyze
 - ```chromosome```: Chromosome number of the variant
 - ```coloc_prob```: colocalization probability threshold
#### summ_file should contain the following columns:
 - ```SNP```: SNP ID in chr:pos:ref:alt format
 - ```allele1```: Effect allele (alt)
 - ```allele2```: Other allele (ref)
 - ```b```: Effect size estimate for effect allele
 - ```se```: Standard error
 - ```N```: Sample size
### Output
The following suffixes are used for output files:
#### Credible Sets:
- GWAS credible sets: ```_credible_set_gwas.txt```
- eQTL credible sets: ```_credible_set_eqtl.txt```
#### Colocalization Results:
-  Summary of colocalization results: ```_coloc_credible_set_summary.txt```
-  Variants within credible sets that reach colocalization probability: ```_coloc_all_SNPs.txt```
## Regional plots for Colocalization
This script generates LocusZoom plots for colocalization analysis between two summary statistics. It supports multiple variants and annotates them together in a single output file.
### Dependencies
Ensure the following R packages are installed:
```
install.packages("data.table")
install.packages("dplyr")
install.packages("locuszoomr")
install.packages("AnnotationHub")
install.packages("stringr")
install.packages("ggplot2")
install.packages("ggrepel")
install.packages("cowplot")
```
### Usage
```
Rscript locuszoom_plot_coloc_SNP.R <disease> "<variant1,variant2,variant3>" "<gene1;gene2>" <index_snp_1> <index_snp_2> <summ_file_1> <summ_file_2> <ld_file_1> <ld_file_2>
```
### Arguments
 - ```disease```: Disease or phenotype name
 - ```variant1,variant2,...```: Comma-separated list of colocalized variants in chr:pos:ref:alt format
 - ```gene1;gene2```: Semicolon-separated list of genes to annotate
- ```index_snp_1```: index SNP for data 1
- ```index_snp_2```: index SNP for data 2
- ```summ_file_1```: summary statistics file 1
- ```summ_file_2```: summary statistics file 2
- ```ld_file_1```: LD matrix file 1
- ```ld_file_2```: LD matrix file 2
### Output
The script generates:
A PDF file ```(<disease>_<variant1>_<variant2>_coloc_SNP.pdf)``` that contains two LocusZoom plots side by side. All input variants are labeled in the same visualization.
